Add wrapper class to make AnalyticSolutions compatible with GH

geoffrey4444 opened the pull request:
## Proposed changes

<!--
At a high level, describe what this PR does.
-->

### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

prayush committed:
Add wrapper class to make AnalyticSolutions compatible with GH

osheamonn commented on src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/WrapGh.hpp:
```diff
@@ -0,0 +1,206 @@
+        get<Tags>(variables(x, t, tmpl::list<Tags>{}, intermediate_vars))...};
+  }
+
+  //  template <typename... Tags>
```
Should this be deleted?

osheamonn commented on src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/WrapGh.hpp:
```diff
@@ -0,0 +1,206 @@
+
+ public:
+  // clang-tidy: google-runtime-references
+  void pup(PUP::er& /*p*/) noexcept {}  // NOLINT
```
This needs to pup the base class by adding `SolutionType::pup(p)`

osheamonn commented on src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/WrapGh.hpp:
```diff
@@ -0,0 +1,206 @@
+template <typename SolutionType>
+inline constexpr bool operator==(const WrapGh<SolutionType>& /*lhs*/,
+                                 const WrapGh<SolutionType>& /*rhs*/) noexcept {
+  return true;
```
This needs to check equality of the base class by casting `lhs` and `rhs` to `SolutionType`. As is, if `lhs` and `rhs` are wrapped KerrSchild solutions with different masses, they will be equal.

osheamonn commented on src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/WrapGh.hpp:
```diff
@@ -0,0 +1,206 @@
+template <typename SolutionType>
+inline constexpr bool operator!=(const WrapGh<SolutionType>& /*lhs*/,
+                                 const WrapGh<SolutionType>& /*rhs*/) noexcept {
+  return false;
```
After changing `operator==` this should be changed to `not(lhs == rhs)`

osheamonn commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Test_WrapGh.cpp:
```diff
@@ -0,0 +1,266 @@
+
+  test_serialization(wrapped_solution);
+  // test operator !=
+  CHECK_FALSE(wrapped_solution != wrapped_solution);
```
I think the test for `operator !=` and `==` should be more rigorous . Maybe construct two KerrSchild solutions with different masses/spins and make sure they're different?

nilsdeppe commented:
Marked as `in progress` since this is waiting on the author.

geoffrey4444 commented:
I'll come back to this in the next couple of days and remove the in progress label once I can respond to the change requests.

geoffrey4444 commented:
I still have to take care of some IWYU issues, but otherwise I think this is ready for another look, since I've made the requested changes

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/WrapGh.hpp:
```diff
@@ -0,0 +1,191 @@
+
+ public:
+  // clang-tidy: google-runtime-references
+  void pup(PUP::er& p) noexcept { SolutionType::pup(p); }  // NOLINT
```
Why not move this to the other `public` section? Our code review guidelines prefer all public before private

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/WrapGh.hpp:
```diff
@@ -0,0 +1,191 @@
+template <typename SolutionType>
+inline constexpr bool operator==(const WrapGh<SolutionType>& lhs,
+                                 const WrapGh<SolutionType>& rhs) noexcept {
+  return static_cast<SolutionType>(lhs) == static_cast<SolutionType>(rhs);
```
`static_cast`s -> `dynamic_cast<const SolutionType&>` since this is runtime polymorphism and we only need to interpret the pointer differently.

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/WrapGh.hpp:
```diff
@@ -0,0 +1,191 @@
+  }
+
+ private:
+  tuples::TaggedTuple<gr::Tags::Lapse<DataVector>> variables(
```
can't pretty much all of these except the GH variables just be inherited? I don't think you'd even need to do `using SolutionType::variables`, but maybe.

geoffrey4444 replied:
Discussed with @nilsdeppe (thanks!) ... I tried this but it didn't compile, and @nilsdeppe is looking to see if there's a nicer solution than just leaving this as is

nilsdeppe replied:
Here:

```cpp
  using TagShift = gr::Tags::Shift<volume_dim, Frame::Inertial, DataVector>;
  using TagSpatialMetric =
      gr::Tags::SpatialMetric<volume_dim, Frame::Inertial, DataVector>;
  using TagInverseSpatialMetric =
      gr::Tags::InverseSpatialMetric<volume_dim, Frame::Inertial, DataVector>;
  using TagExCurvature =
      gr::Tags::ExtrinsicCurvature<volume_dim, Frame::Inertial, DataVector>;

#define FUNC_DECL(r, data, elem)                                            \
  tuples::TaggedTuple<elem> variables(                                      \
      const tnsr::I<DataVector, volume_dim>& /*x*/, double /*t*/,           \
      tmpl::list<elem> /*meta*/, const IntermediateVars& intermediate_vars) \
      const noexcept;

#define MY_LIST                                                              \
  BOOST_PP_TUPLE_TO_LIST(                                                    \
      11,                                                                    \
      (gr::Tags::Lapse<DataVector>, TimeDerivLapse, DerivLapse, TagShift,    \
       TimeDerivShift, DerivShift, TagSpatialMetric, TimeDerivSpatialMetric, \
       DerivSpatialMetric, TagInverseSpatialMetric, TagExCurvature,          \
       gr::Tags::SqrtDetSpatialMetric<DataVector>))

  BOOST_PP_LIST_FOR_EACH(FUNC_DECL, _, MY_LIST)
#undef MY_LIST
#undef FUNC_DECL
```
and you need the includes:
```cpp
#include <boost/preprocessor/list/for_each.hpp>
#include <boost/preprocessor/tuple/to_list.hpp>
```

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/WrapGh.hpp:
```diff
@@ -0,0 +1,191 @@
+ * and `GeneralizedHarmonic::Tags::Phi`
+ */
+template <typename SolutionType>
+class WrapGh : public SolutionType {
```
I'm confused by the name. This is a `Gh` solution, it wraps a `GR` solution, so why not `WrapGr` (or better yet `WrappedGr`)?

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Test_WrapGh.cpp:
```diff
@@ -0,0 +1,286 @@
+
+template <typename SolutionType>
+void test_generalized_harmonic_solution(
+    const SolutionType& solution,
```
Why not make this a variadic function as follows:
```cpp
template <typename SolutionType>
void test_generalized_harmonic_solution(const Args&... args) noexcept {
  SolutionType solution{args...};
  GeneralizedHarmonic::Solutions::WrapGh<SolutionType> wrapped_solution{args...};
  // Other testing code.

  // Test serialization:
  test_serialization(wrapped_solution);

  test_copy_and_move(wrapped_solution);
}
```

Then the test case is just:
```cpp
SPECTRE_TEST_CASE("Unit.PointwiseFunctions.AnalyticSolutions.Gr.WrapGh",
                  "[PointwiseFunctions][Unit]") {
  test_generalized_harmonic_solution<gr::Solutions::Minkowski<1>>();
  test_generalized_harmonic_solution<gr::Solutions::Minkowski<2>>();
  test_generalized_harmonic_solution<gr::Solutions::Minkowski<3>>();

  const double mass = 0.5;
  const std::array<double, 3> spin{{0.1, 0.2, 0.3}};
  const std::array<double, 3> center{{1.0, 3.0, 2.0}};
  test_generalized_harmonic_solution<gr::Solutions::KerrSchild>(mass, spin, center);
  
  const double mass2 = 0.4;
  const std::array<double, 3> spin2{{0.4, 0.3, 0.2}};
  const std::array<double, 3> center2{{4.0, 1.0, 3.0}};
  test_generalized_harmonic_solution<gr::Solutions::KerrSchild>(mass2, spin2, center2);

  test_construct_from_options();
}
```

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Test_WrapGh.cpp:
```diff
@@ -0,0 +1,286 @@
+  // Get quantities from an analytic solution of Einstein's equations
+  const auto vars = solution.variables(
+      x, t, typename SolutionType::template tags<DataVector>{});
+  const auto& lapse = get<gr::Tags::Lapse<DataVector>>(vars);
```
All of these non-GH checks can be done as a for loop:
```cpp
tmpl::for_each<typename SolutionType::template tags<DataVector>>(
  [&vars, &wrapped_vars](auto tag_v) noexcept {
    using tag = typename decltype(tag_v)::type;
    CHECK(get<tag>(vars) == get<tag>(wrapped_vars));
  });
```

nilsdeppe reviewed (changes requested):
Since I'm basically asking you to condense 50% of the code in this PR I'm happy for you to squash :)

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/WrappedGr.cpp:
```diff
@@ -0,0 +1,438 @@
+/// \cond
+namespace GeneralizedHarmonic {
+namespace Solutions {
+template <typename SolutionType>
```
@geoffrey4444 here you need:
```cpp
#define FUNC_DEF(r, data, elem)                                            \
  template <typename SolutionType>                                         \
  auto WrappedGr<SolutionType>::variables(                                 \
      const tnsr::I<DataVector, GeneralizedHarmonic::Solutions::WrappedGr< \
                                    SolutionType>::volume_dim>& /*x*/,     \
      double /*t*/, tmpl::list<elem> /*meta*/,                             \
      const IntermediateVars& intermediate_vars)                           \
      const noexcept->tuples::TaggedTuple<elem> {                          \
    return {get<elem>(intermediate_vars)};                                 \
  }

#define MY_LIST                                                              \
  BOOST_PP_TUPLE_TO_LIST(                                                    \
      11,                                                                    \
      (gr::Tags::Lapse<DataVector>, TimeDerivLapse, DerivLapse, TagShift,    \
       TimeDerivShift, DerivShift, TagSpatialMetric, TimeDerivSpatialMetric, \
       DerivSpatialMetric, TagInverseSpatialMetric, TagExCurvature,          \
       gr::Tags::SqrtDetSpatialMetric<DataVector>))

  BOOST_PP_LIST_FOR_EACH(FUNC_DEF, _, MY_LIST)
#undef MY_LIST
#undef FUNC_DEF
```
I was only able to get this working with the preprocessor, not anything else :man_shrugging:

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/WrappedGr.hpp:
```diff
@@ -0,0 +1,191 @@
+  void pup(PUP::er& p) noexcept { SolutionType::pup(p); }  // NOLINT
+
+ private:
+  tuples::TaggedTuple<gr::Tags::Lapse<DataVector>> variables(
```
You'll need includes:
```cpp
#include <boost/preprocessor/list/for_each.hpp>
#include <boost/preprocessor/tuple/to_list.hpp>
```

and can replace the function decls with:
```cpp
  using TagShift = gr::Tags::Shift<volume_dim, Frame::Inertial, DataVector>;
  using TagSpatialMetric =
      gr::Tags::SpatialMetric<volume_dim, Frame::Inertial, DataVector>;
  using TagInverseSpatialMetric =
      gr::Tags::InverseSpatialMetric<volume_dim, Frame::Inertial, DataVector>;
  using TagExCurvature =
      gr::Tags::ExtrinsicCurvature<volume_dim, Frame::Inertial, DataVector>;

#define FUNC_DECL(r, data, elem)                                            \
  tuples::TaggedTuple<elem> variables(                                      \
      const tnsr::I<DataVector, volume_dim>& /*x*/, double /*t*/,           \
      tmpl::list<elem> /*meta*/, const IntermediateVars& intermediate_vars) \
      const noexcept;

#define MY_LIST                                                              \
  BOOST_PP_TUPLE_TO_LIST(                                                    \
      11,                                                                    \
      (gr::Tags::Lapse<DataVector>, TimeDerivLapse, DerivLapse, TagShift,    \
       TimeDerivShift, DerivShift, TagSpatialMetric, TimeDerivSpatialMetric, \
       DerivSpatialMetric, TagInverseSpatialMetric, TagExCurvature,          \
       gr::Tags::SqrtDetSpatialMetric<DataVector>))

  BOOST_PP_LIST_FOR_EACH(FUNC_DECL, _, MY_LIST)
#undef MY_LIST
#undef FUNC_DECL
```

nilsdeppe reviewed (commented):
Okay, got it working with preprocessor magic :) I'm happy for you to squash :)

geoffrey4444 commented:
Thanks for the solution @nilsdeppe! I implemented it and squashed. I think this is ready for another look when you get a chance, @osheamonn and @nilsdeppe

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/WrappedGr.cpp:
```diff
@@ -0,0 +1,304 @@
+namespace Solutions {
+// Preprocessor logic to avoid defining variables() functions for
+// tags other than the three the wrapper adds (i.e., other than
+// gr::Tags::SpacetimeMetric, gr::Tags::Lapse, and gr::Tags::Shift.
```
Shouldn't `Lapse` and `Shift` be `GeneralizedHarmonic::Tags::Phi` and `GeneralizedHarmonic::Tags::Pi`? And you're missing a closing paren :)

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/WrappedGr.hpp:
```diff
@@ -0,0 +1,156 @@
+ private:
+  // Preprocessor logic to avoid declaring variables() functions for
+  // tags other than the three the wrapper adds (i.e., other than
+  // gr::Tags::SpacetimeMetric, gr::Tags::Lapse, and gr::Tags::Shift.
```
Shouldn't `Lapse` and `Shift` be `GeneralizedHarmonic::Tags::Phi` and `GeneralizedHarmonic::Tags::Pi`? And you're missing a closing paren :)

geoffrey4444 replied:
Yes, my brain made a caching error :) Fixed and squashed

nilsdeppe reviewed (commented):
Just some slight confusion on a comment. This looks good :)

nilsdeppe reviewed (dismissed):
LGTM :) This was a really good idea on your and Prayush's part!

geoffrey4444 commented:
Fixed iwyu and squashed

geoffrey4444 commented:
When you get a chance, could @osheamonn please take a look?

geoffrey4444 commented:
@osheamonn I have a long flight tomorrow, so if you can take a look at the WrappedGr PR today, I can respond to your comments on the plane

kidder commented:
@geoffrey4444 on your long flight, review all the PRs on the first page 😛

geoffrey4444 commented:
I’ll cache the pages and see if I can do this without internet, submitting when I get back. No promises about getting to all of them, but I’ll try to do some of them!

osheamonn commented on src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/WrappedGr.hpp:
```diff
@@ -0,0 +1,157 @@
+template <typename SolutionType>
+inline constexpr bool operator==(const WrappedGr<SolutionType>& lhs,
+                                 const WrappedGr<SolutionType>& rhs) noexcept {
+  return dynamic_cast<const SolutionType&>(lhs) ==
```
My only question is why these are dynamic_casts? @nilsdeppe said to use dynamic instead of static since this is runtime polymorphism, but my understanding was that this class is used for static polymorphism. In the GH executable PR, the solution type used is a `WrapGh<KerrSchild>`, which seems like static polymorphism to me. Maybe I've misunderstood how this class is to be used or how the runtime polymorphism ends up working out?

nilsdeppe replied:
Either is generally fine in this case. My reason for preferring a `dynamic_cast` is that with it at least if there is a bug introduce in the code you will get a runtime error from the cast rather than UB. Since we can reasonably make the guarantee that `WrapGh<B>` has `B` as a base class a static cast would be fine. Here it's a matter of personal preference and not having to worry about the performance overhead that the runtime check in a dynamic cast adds since we shouldn't be comparing analytic solutions for equality all that often.
